src/solvers/petsc_nonlinear_solver.C

```diff
@@ -182,10 +182,19 @@ extern "C"
     else
       libmesh_error_msg("Error! Unable to compute residual and/or Jacobian!");
 
-    PetscVector<Number> X(x, rc.sys.comm());
+
+    // Synchronize PETSc x to local solution since the local solution may be changed due to the constraints
+    PetscVector<Number> & X_sys = *cast_ptr<PetscVector<Number> *>(rc.sys.solution.get());
+    PetscVector<Number> X_global(x, rc.sys.comm());
+
+    X_global.swap(X_sys);
+    rc.sys.update();
+    X_global.swap(X_sys);
```

Comment:
This is interesting. So we `enforce_constraints_exactly` on the `current_local_solution` before the residual evaluation. But here we are restoring the `current_local_solution` to PETSc `x` so that when perturbing `x` we will actually see those effects appear when we `enforce_constraints_on_residual` post residual evaluation. I'm down for this. I assume that you've checked the performance on this and it looks good?